server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java

```diff
@@ -0,0 +1,146 @@
+   }
+
+   /**
+    * This reserves the current thread's executor so that no other executor may submit tasks to it including itself.
+    * This should be used if the calling code will end up blocking when it shouldn't and is a sign of code that
+    * should eventually be refactored but provides a way to share the netty event loop
+    */
+   public static void reserveCurrentThread() {
+      EventExecutor thisExecutor = ThreadExecutorMap.currentExecutor();
+      if (thisExecutor != null) {
+         boolean reserved = reserveExecutors.add(thisExecutor);
+         assert reserved;
+      }
+   }
+
+   /**
+    * No longer reserves the current thread and allows it to be submitted to. This should always be invoked when
+    * the blocking in the invoking thread is complete as soon as possible.
+    */
+   public static void unreserveCurrentThread() {
+      EventExecutor thisExecutor = ThreadExecutorMap.currentExecutor();
+      if (thisExecutor != null) {
+         boolean unreserved = reserveExecutors.remove(thisExecutor);
+         assert unreserved;
+      }
+   }
+
+   private EventExecutor getExecutorNotInEventLoop() {
+      for (int i = 0; i < loopAttempts; ++i) {
+         EventExecutor eventExecutor = eventLoopGroup.next();
+         // We don't want to submit a task to our current thread as some callers may block waiting for it to complete.
+         if (eventExecutor.inEventLoop()) {
+            log.tracef("Skipped submitting task to %s as it is the current event loop - trying another", eventExecutor);
+            continue;
+         }
+         // We also don't want to submit to a reserved executor as it is currently blocking waiting for another caller
+         if (reserveExecutors.contains(eventExecutor)) {
+            log.tracef("Skipped submitting task to %s as it is currently reserved - trying another", eventExecutor);
+            continue;
+         }
+
+         return eventExecutor;
+      }
+      log.fatalf("No executor was found - replaying to trace issue - current reserveExecutors are %s", reserveExecutors);
+      for (int i = 0; i < loopAttempts; ++i) {
+         EventExecutor eventExecutor = eventLoopGroup.next();
+         // We don't want to submit a task to our current thread as some callers may block waiting for it to complete.
+         if (eventExecutor.inEventLoop()) {
+            log.fatalf("Skipped submitting task to %s as it is the current event loop - trying another", eventExecutor);
+            continue;
+         }
+         // We also don't want to submit to a reserved executor as it is currently blocking waiting for another caller
+         if (reserveExecutors.contains(eventExecutor)) {
+            log.fatalf("Skipped submitting task to %s as it is currently reserved - trying another", eventExecutor);
+            continue;
+         }
+
+         log.fatalf("Found a matched executor %s, possibly concurrent reservation?", eventExecutor);
```

Comment:
can't you return the `eventExecutor` instead of throwing an `IllegalStateException`

Reply:
I don't want to. This was a sign that there was a bug, I would rather find it sooner rather than later.

Reply:
> This was a sign that there was a bug

not really, it could be just bad luck

It looks like `next()` is not predictable (i.e. it is a shared counter, so requests from other threads will affect the return value in this loop) so you may never test all the executors in your loop. (worst case, it can return the same executor `loopAttempts` times)

Reply:
No, it is definitely a bug. If this is ever thrown that means it cannot use an executor. All the implementations of the netty event loop chooser by default will touch each event once in a loop of the same size. They use a modulus or power of two by incrementing a value. If they ever changed the implementation of the event executor chooser, then I agree it could be luck (which is one reason I loop through total number times two just in case). But as it is now it is very deterministic.

Reply:
maybe I'm missing something... 
It looks like this code https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/DefaultEventExecutorChooserFactory.java#L73 is invoked when `next()` is invoked. 

If you have 2 threads submitting a task, one thread could get the executors in the even position and the other in the odd as an example. I don't see any code that suggests that all executors will be touched.

Reply:
I was referring to a unit test which would be hard pressed to run into this. But actually this reminds me why I had the replay logic with tracing so it would expose if this ever did happen :) Since the message you linked would be printed in such a case.

By the  way I forced it to two threads for runs in radargun and multiple times in CI and hasn't run across this yet :)

But this just reinforces using the reserve as little as possible.

Reply:
But thinking on it more I don't know if I am being critical enough for the concurrency case. Let me fix up the config ones at least and I will take a peak at the others more closely. The retrieving cache and counter cache ones should be extremely  rare and tasks currently are as well. But I will see if we can reduce this further.

Reply:
I can provide a custom event executor chooser which should remove the randomness and the only way to get the exception would be if all threads are reserved.

Reply:
Then the only way to run into the issue is having N concurrent calls to `reserveCurrentThread` where N is the number of event loops in the system.